src/modules/CRUD/services/CRUDProvider.js

```diff
@@ -29,105 +33,185 @@ export class CRUDProvider extends React.Component {
+
+    this.setState({
+      featuresList,
+      errors: {
+        ...errors,
+        featuresList: this.getFormattedError({ error, ids: { layerId }, store: 'featuresList' }),
+      },
+    });
+    return featuresList.length && featuresList;
+  };
 
   fetchFeature = async (layerId, featureId) => {
+    const result = {};
     try {
       const feature = await fetchFeatureAction(layerId, featureId);
-      this.setState(state => ({
-        feature: {
-          ...state.feature,
-          [feature.identifier]: feature,
-        },
-      }));
+      result.feature = feature;
     } catch (e) {
-      this.setState(state => ({
-        ...state,
-        errors: { ...state.errors, [featureId]: true, code: e.message },
-      }));
+      result.error = e;
     }
+
+    const { feature = {}, error = {} } = result;
+    const { featuresList: prevFeaturesList } = this.state;
+    const isFeatureAlreadyExisting = prevFeaturesList.some(({ identifier }) => (
+      identifier === feature.identifier
+    ));
+
+
+    const featuresList = isFeatureAlreadyExisting
+      ? prevFeaturesList.map(item => (item.identifier === feature.identifier ? feature : item))
+      : [...prevFeaturesList, feature.identifier && feature].filter(Boolean);
+
+    this.setState(({ errors }) => ({
+      featuresList,
+      errors: {
+        ...errors,
+        feature: this.getFormattedError({ error, ids: { layerId, featureId }, store: 'feature' }),
+      },
+    }));
+
+    return Object.keys(feature).length && feature;
   }
 
   saveFeature = async (layerId, featureId, data) => {
+    const result = {};
     try {
       const feature = await saveFeatureAction(layerId, featureId, data);
-      const { featuresList } = this.state;
-      const featureAlreadyExists = featuresList.some(({ identifier }) =>
-        identifier === feature.identifier);
-
-      this.setState(state => ({
-        feature: {
-          ...state.feature,
-          [feature.identifier]: feature,
-        },
-        featuresList: featureAlreadyExists
-          ? featuresList.map(item => (item.identifier === feature.identifier ? feature : item))
-          : [...featuresList, feature],
-      }));
-      return feature;
+      result.feature = feature;
     } catch (e) {
-      this.setState(state => ({
-        ...state,
-        errors: { ...state.errors, [featureId]: e.message },
-      }));
-      return null;
+      result.error = e;
     }
+
+    const { feature = {}, error = {} } = result;
+
+    const { featuresList: prevFeaturesList } = this.state;
+
+    const isFeatureAlreadyExisting = prevFeaturesList.some(({ identifier }) => (
+      identifier === feature.identifier
+    ));
+
+    const featuresList = isFeatureAlreadyExisting
+      ? prevFeaturesList.map(item => (item.identifier === feature.identifier ? feature : item))
+      : [...prevFeaturesList, (feature.identifier) && feature].filter(Boolean);
+
+    this.setState(({ errors }) => ({
+      featuresList,
```

Comment:
`(prev)featuresList` should also be read from `setState` callback argument (like `error`), and not from a previous `this.state`.

Maybe `featuresList` should be defined using pure functions getting `prevFeaturesList` and `feature` as argument:

```js
this.setState(({  errors, featureList: prevFeatureList }) => ({
  featureList: onlyExistingFeatures(prevFeatureList, feature),
  errors: {
  // ...
```